distributed/deploy/adaptive.py

```diff
@@ -28,49 +33,154 @@ class Adaptive(object):
+        total_cores = sum(self.scheduler.ncores.values())
+
+        if total_occupancy / (total_cores + 1e-9) > self.startup_cost * 2:
+            logger.info("CPU limit exceeded [%d occupancy / %d cores]",
+                        total_occupancy, total_cores)
+            return True
+        else:
+            return False
+
+    def needs_memory(self):
+        """
+        Check if the cluster is RAM constrained
+
+        Notes
+        -----
+        Returns ``True`` if  the required bytes in distributed memory is some
+        factor larger than the actual distributed memory available.
+        """
+        limit_bytes = {w: self.scheduler.worker_info[w]['memory_limit']
+                        for w in self.scheduler.worker_info}
+        worker_bytes = self.scheduler.worker_bytes
+
+        limit = sum(limit_bytes.values())
+        total = sum(worker_bytes.values())
+        if total > 0.6 * limit:
+            logger.info("Ram limit exceeded [%d/%d]", limit, total)
+            return True
+        else:
+            return False
+
     def should_scale_up(self):
+        """
+        Determine whether additional workers should be added to the cluster
+
+        Returns
+        -------
+        scale_up : bool
+
+        Notes
+        ----
+        Additional workers are added whenever
+
+        1. There are unrunnable tasks and no workers
+        2. The cluster is CPU constrained
+        3 The cluster is RAM constrained
+
+        See Also
+        --------
+        needs_cpu
+        needs_memory
+        """
         with log_errors():
             if self.scheduler.unrunnable and not self.scheduler.ncores:
                 return True
 
-            total_occupancy = sum(self.scheduler.occupancy.values())
-            total_cores = sum(self.scheduler.ncores.values())
+            needs_cpu = self.needs_cpu()
+            needs_memory = self.needs_memory()
 
-            if total_occupancy / (total_cores + 1e-9) > self.startup_cost * 2:
+            if needs_cpu or needs_memory:
                 return True
 
-            limit_bytes = {w: self.scheduler.worker_info[w]['memory_limit']
-                            for w in self.scheduler.worker_info}
-            worker_bytes = self.scheduler.worker_bytes
+            return False
 
-            limit = sum(limit_bytes.values())
-            total = sum(worker_bytes.values())
+    def should_scale_down(self):
+        """
+        Determine whether any workers should potentially be removed from
+        the cluster.
 
-            if total > 0.6 * limit:
-                return True
+        Returns
+        -------
+        scale_down : bool
 
-            return False
+        Notes
+        -----
+        ``Adaptive.should_scale_down`` always returns True, so we will always
+        attempt to remove workers as determined by
+        ``Scheduler.workers_to_close``.
+
+        See Also
+        --------
+        Scheduler.workers_to_close
+        """
+        return True
```

Comment:
Agreed :) My usecase is allowing a subclass to define something more useful. For example, an Adaptive strategy that won't scale down too soon after the last task finished:

```python
import datetime
from distributed.diagnostics.plugin import SchedulerPlugin
from distributed import LocalCluster, Client
from distributed.deploy.adaptive import Adaptive


class LastTask(SchedulerPlugin):
    "Record the time of the last task"
    def __init__(self):
        self.last_task = datetime.datetime.now()

    def transition(self, key, start, finish):
        if start == 'waiting' and finish == 'processing':
            self.last_task = datetime.datetime.now()


plugin = LastTask()
cluster = LocalCluster(n_workers=0)
client = Client(cluster)
cluster.scheduler.add_plugin(plugin)


class DownsizeSlowly(Adaptive):

    def should_scale_down(self):
        delta = datetime.datetime.now() - plugin.last_task
        if delta > datetime.timedelta(seconds=5):
            print("Scaling down")
            return True
        else:
            print("too soon: ", delta)


adapt = DownsizeSlowly(cluster.scheduler, cluster)


f = client.submit(lambda x: x + 1, 10)
```

Do you have a more useful default in mind? Or a better way override the default behavior? I suppose an alternative is to have all the scale-down logic on a separate `interval`, though I could see that getting complicated.